borsh-derive/src/internals/serialize/enums/mod.rs

```diff
@@ -41,18 +44,44 @@ pub fn process(input: &ItemEnum, cratename: Path) -> syn::Result<TokenStream2> {
         }
     }
     let fields_body = optimize_fields_body(fields_body, has_unit_variant);
-    generics_output.extend(&mut where_clause, &cratename);
+    generics_output.extend::<IS_ASYNC>(&mut where_clause, &cratename);
+
+    let serialize_trait = Ident::new(
+        if IS_ASYNC {
+            "BorshSerializeAsync"
+        } else {
+            "BorshSerialize"
+        },
+        Span::call_site(),
+    );
+    let writer_trait = if IS_ASYNC {
+        quote! { async_io::AsyncWrite }
+    } else {
+        quote! { io::Write }
+    };
+    let r#async = IS_ASYNC.then(|| Token![async](Span::call_site()));
+    let lifetime = IS_ASYNC.then(|| Lifetime::new("'async_variant", Span::call_site()));
+    let lt_comma = IS_ASYNC.then(|| Token![,](Span::call_site()));
+
+    let write_variant_idx = if IS_ASYNC {
+        quote! { writer.write_u8(variant_idx).await }
+    } else {
+        quote! { writer.write_all(&variant_idx.to_le_bytes()) }
+    };
```

Comment:
issue: plz create a todo issue to replace these with `BorshSerialize(Async)::serialize` (not in the scope of this pr)